lhotse/recipes/ami.py

```diff
@@ -0,0 +1,256 @@
+        audio_name = re.sub(r'\..*$', '', audio_idx)
+        if audio_idx not in anotation_lists:
+            logging.warning(f'No annotation found for {audio_idx}')
+            continue
+        audio_info = torchaudio.info(str(audio_path))[0]
+
+        # Split the raw data into segments
+        anotation = anotation_lists[audio_idx]
+        if asr_task:
+            # For ASR task, we split the audio not only by silences, but also by punctuations
+            for seg_idx, seg_info in enumerate(anotation):
+                for subseg_idx, subseg_info in enumerate(seg_info):
+                    duration = subseg_info.end_time-subseg_info.begin_time
+                    if duration > 0:
+                        metadata[f'{audio_idx}-{seg_idx}-{subseg_idx}'] = AmiMetaData(
+                            audio_path=audio_path,
+                            audio_name=audio_name,
+                            audio_info=audio_info,
+                            text=subseg_info.text,
+                            offset_seconds=subseg_info.begin_time,
+                            duration_seconds=duration
+                        )
+        else:
+            # For VAD task, we simply split the audio by silences
+            current_time = 0.0
+            seg_idx = 0
+            for seg_info in anotation:
+                begin_time = seg_info[0].begin_time
+                end_time = seg_info[-1].end_time
+                if current_time < begin_time:
+                    # Insert a unvoiced segment
+                    metadata[f'{audio_idx}-{seg_idx}u'] = AmiMetaData(
+                        audio_path=audio_path,
+                        audio_name=audio_name,
+                        audio_info=audio_info,
+                        text='',
+                        offset_seconds=current_time,
+                        duration_seconds=(begin_time-current_time)
+                    )
+                    seg_idx += 1
+                # Then the current segment
+                metadata[f'{audio_idx}-{seg_idx}v'] = AmiMetaData(
+                    audio_path=audio_path,
+                    audio_name=audio_name,
+                    audio_info=audio_info,
+                    text='[voice]',
+                    offset_seconds=begin_time,
+                    duration_seconds=(end_time - begin_time)
+                )
+                seg_idx += 1
+                current_time = end_time
+
+    return metadata
+
+
+def prepare_ami(
+        data_dir: Pathlike,
+        output_dir: Pathlike,
+        asr_task: Optional[bool] = True,
+        write_yaml: Optional[bool] = False,
+) -> Dict[str, Dict[str, Union[RecordingSet, SupervisionSet]]]:
+    manifests = defaultdict(dict)
+
+    data_dir = Path(data_dir)
+    output_dir = Path(output_dir)
+    output_dir.mkdir(parents=True, exist_ok=True)
+
+    metadata = make_metadata(data_dir, asr_task)
+    for part in dataset_parts:
+        # Audio
+        audio = RecordingSet.from_recordings(
```

Comment:
Hmm, this doesn't seem right to me. The idea behind `Recording` is that even if a wave file is 10 hours long, it's a single recording. It just doesn't care about the segmentation (like `wav.scp` in Kaldi). So for each audio file, you should have one `Recording`, and many different `SupervisionSegment`s, that refer to this recording via `recording_id`.

Then, you don't need the `offset_seconds` field that you added to `Recording`. The `FeatureSetBuilder` is going to extract the features for the whole duration of the recording. After that, you can create the `CutSet` with the method named like `make_windowed_cuts_...()` with an even duration of e.g. 10 seconds. I don't remember if it handles supervisions, but if not, then you'd make it preserve the supervisions (just cut their time boundaries to not exceed those of the cut, I think this is already implemented when you call `cut.truncate()`). As a result, you'll get cuts that have some frames with speech and some frames without, with that information encoded by the presence of `SupervisionSegment`s. 

If you do it this way, you also won't need the `asr_task` flag.

Reply:
Actually the best way to implement the windowing would be to add a new method that creates a cut set, that takes a feature set and a supervision set, and creates one cut per each feature (not each supervision, like the current one) ((that method could actually be instead a flag to the `from_manifests()`, maybe called `cut_on_supervisions: bool = True`)). Then, to create evenly sized cuts, you'd just iteratively call `cut.truncate(offset, offset + duration, keep_excessive_supervisions=True)`, and it's going to do the right thing with the supervisions for you. This could maybe be another method on `CutSet`, named e.g. `split_cuts`?

Reply:
Piotr, I don't quite understand what you are saying here, e.g. "creates one cut per each feature".

Reply:
... also, my recollection is that AMI is quite complicated w.r.t. microphones/channels, e.g. we have headset microphones and multiple distant microphones, and in Kaldi we have several different ways of dealing with that.  Can someone please clarify the approach that is being taken here?  This should probably be documented, e.g. in a docstring at the top of the file or wherever is appropriate.

Reply:
I agree with the comment about AMI. Regarding the "cut per each feature", please check #65 - I drafted the changes that I previously described and explained a bit more.